### PR TITLE
CSRF attack should not save answers on page

### PR DESCRIPTION
**Pull request by benjefferies**
### What is the context of this PR?
[trello](https://trello.com/c/VLUhPJEE/1006-csrf-behaviour)

If a csrf token is invalid log the user out and do not save the answers.

### How to review 
Tests in test_questionnaire_csrf.py demonstrate how it works.

To manually test
1. Open a questionnaire
1. Open chrome developer console
1. Search for csrf_token
1. Change csrf_token
1. Submit page

Expect to be signed out

### CODE REVIEW

**Comment by jonnyshaw89** [comment]
@benjefferies Do we want to log the user out?

File: tests/integration/questionnaire/test_questionnaire_csrf.py
Comment (collisdigital) [inline tests/integration/questionnaire/test_questionnaire_csrf.py]: :-1: for tests in existing format; I've seen the light!

Comment (collisdigital) [reply tests/integration/questionnaire/test_questionnaire_csrf.py]: Not that there's anything to be done until the integration test PR merges; I'll re-base (the other PR) and update this when it merges to the new format.

**Comment by collisdigital** [comment]
As a general service comment I think we need to prioritise the work to put the log me back in on the signed-out page as we keep adding to the number of ways a user can get logged out with now easy way for them to get back in

**Comment by collisdigital** [comment]
I've commented on this https://trello.com/c/CDmNd8hd/901-respondent-journey

**Comment by benjefferies** [comment]
I agree it will be good to review the user journey to reflect on how it currently is and talk through approaches of improving it. 

It's worth baring in mind that this scenario will be the minority as you'd not imagine there would be many users targeted with csrf but if one does it would be unclear at the moment of what went wrong and how to recover.

**Comment by collisdigital** [comment]
I recommend we ignore the POST if invalid and re-direct to a 403 Forbidden on the current URL; if they refresh they'll recover back to the empty/last state of the page. This also means if they have an existing tab/window open that /is/ logged in they can continue there without having to sign-in again. 

As part of the Respondent Journey the error pages can be reviewed.

**Comment by collisdigital** [comment]
I've added A/C to https://trello.com/c/VLUhPJEE/1006-csrf-behaviour

**Comment by benjefferies** [comment]
It's worth discussing that point with the security team as Dan had suggested logging the user out. Not sure if he's identified a session hijacking scenario that the log out suggestion was to protect. His suggestion was to: `Options: Destroy the session; display an error age from which the valid user cannot progress.`

**Comment by collisdigital** [comment]
I've also added some scenarios to the card; really this card shouldn't have come into development until these were agreed...

**Comment by collisdigital** [comment]
I've discussed with Dan, he's happy with the scenarios I've added.

File: tests/integration/questionnaire/test_questionnaire_csrf.py
Comment (collisdigital) [inline tests/integration/questionnaire/test_questionnaire_csrf.py:61]: Any reason to use a different survey; in that this overrides what happens in `setUp` above?

File: tests/integration/questionnaire/test_questionnaire_csrf.py
Comment (collisdigital) [inline tests/integration/questionnaire/test_questionnaire_csrf.py]: I had been in-lining these to keep the verbosity down e.g.

`self.post({'favourite-breakfast': 'Muesli'})`

File: tests/integration/questionnaire/test_questionnaire_csrf.py
Comment (collisdigital) [inline tests/integration/questionnaire/test_questionnaire_csrf.py]: There is a helper for this `self.getResponseData()`

File: tests/integration/questionnaire/test_questionnaire_csrf.py
Comment (collisdigital) [inline tests/integration/questionnaire/test_questionnaire_csrf.py]: If we can avoid putting URLs in it makes the tests more resilient to changes in future, if you use the survey launched in `setUp()` you can also use `self.first_url` or record a URL after a post/get from `self.last_url` to re-submit to the same page.

File: tests/integration/questionnaire/test_questionnaire_csrf.py
Comment (collisdigital) [inline tests/integration/questionnaire/test_questionnaire_csrf.py]: There is a helper for this `self.getResponseData()` wonder if making last_response private is better (not sure if any other tests access it directly)

File: tests/integration/questionnaire/test_questionnaire_csrf.py
Comment (collisdigital) [inline tests/integration/questionnaire/test_questionnaire_csrf.py]: you can pass the action to `post()` e.g.

`self.post({'favourite-breakfast': 'Muesli'}, action='save_sign_out')`

File: tests/integration/questionnaire/test_questionnaire_csrf.py
Comment (collisdigital) [inline tests/integration/questionnaire/test_questionnaire_csrf.py]: Here too

File: tests/integration/questionnaire/test_questionnaire_csrf.py
Comment (collisdigital) [inline tests/integration/questionnaire/test_questionnaire_csrf.py]: align with other tests; use `self.last_csrf_token`

File: tests/integration/questionnaire/test_questionnaire_csrf.py
Comment (collisdigital) [inline tests/integration/questionnaire/test_questionnaire_csrf.py]: Here too

File: tests/integration/questionnaire/test_questionnaire_csrf.py
Comment (benjefferies) [reply tests/integration/questionnaire/test_questionnaire_csrf.py]: Done

File: tests/integration/questionnaire/test_questionnaire_csrf.py
Comment (benjefferies) [reply tests/integration/questionnaire/test_questionnaire_csrf.py]: Done

File: tests/integration/questionnaire/test_questionnaire_csrf.py
Comment (benjefferies) [reply tests/integration/questionnaire/test_questionnaire_csrf.py:61]: The household composition question has a bespoke end point so ensuring there isn't a csrf bug or regression there

File: tests/integration/questionnaire/test_questionnaire_csrf.py
Comment (benjefferies) [reply tests/integration/questionnaire/test_questionnaire_csrf.py]: This page is a few blocks into the questionnaire so don't think it's worth progressing through the questionnaire just to post to this question

File: tests/integration/questionnaire/test_questionnaire_csrf.py
Comment (benjefferies) [reply tests/integration/questionnaire/test_questionnaire_csrf.py]: Done

File: tests/integration/questionnaire/test_questionnaire_csrf.py
Comment (benjefferies) [reply tests/integration/questionnaire/test_questionnaire_csrf.py]: Done

File: tests/integration/questionnaire/test_questionnaire_csrf.py
Comment (benjefferies) [reply tests/integration/questionnaire/test_questionnaire_csrf.py]: Done

File: tests/integration/questionnaire/test_questionnaire_csrf.py
Comment (collisdigital) [inline tests/integration/questionnaire/test_questionnaire_csrf.py]: Would be safer to use a value other than Pancakes here?

Comment (benjefferies) [reply tests/integration/questionnaire/test_questionnaire_csrf.py]: Yeah think it was muesli and pancake but maybe copy and paste errror

File: tests/integration/questionnaire/test_questionnaire_csrf.py
Comment (collisdigital) [reply tests/integration/questionnaire/test_questionnaire_csrf.py:61]: Maybe put this in a separate test class then as this test is doing a load of stuff in setUp and the overwriting it, I guess that's not a major issue just seems wrong

Comment (benjefferies) [reply tests/integration/questionnaire/test_questionnaire_csrf.py:61]: Not sure if the set up is really necessary as each test could do it's own set up. I think the introduction check is probably redundant as it would be annoying if all tests fail for an unrelated issue. Also I don't think repetitiveness in tests are that bad and sometimes can help with the readability of the scenario as each test is essentially the scenario

Comment (collisdigital) [reply tests/integration/questionnaire/test_questionnaire_csrf.py:61]: Agreed, ditch the setUp

Comment (collisdigital) [reply tests/integration/questionnaire/test_questionnaire_csrf.py:61]: Also only some tests need the dumper roles if you remove the common setup

File: tests/integration/questionnaire/test_questionnaire_csrf.py
Comment (jonnyshaw89) [inline tests/integration/questionnaire/test_questionnaire_csrf.py]: This needs to be updated from the copy/paste?

File: app/views/questionnaire.py
Comment (jonnyshaw89) [inline app/views/questionnaire.py:386]: This exception doesn't appear to be caught by our errors view to render our happy error pages.

**Review by jonnyshaw89 — CHANGES_REQUESTED** [review]

File: tests/integration/questionnaire/test_questionnaire_csrf.py
Comment (benjefferies) [reply tests/integration/questionnaire/test_questionnaire_csrf.py:61]: Done

File: tests/integration/questionnaire/test_questionnaire_csrf.py
Comment (benjefferies) [reply tests/integration/questionnaire/test_questionnaire_csrf.py]: Done

File: app/views/questionnaire.py
Comment (benjefferies) [reply app/views/questionnaire.py:386]: Good spot, done

**Review by collisdigital — APPROVED** [review]

**Review by jonnyshaw89 — APPROVED** [review]

**Comment by collisdigital** [comment]
Fix failing tests and squash, then it's ready.

**Pull request merged**
